README
Introduction
============

This document aims to explain the Directory Information Tree (DIT) used in the
openldap-dit package.

The motivation for this new layout is the need for a better separation of
privileges regarding access to the information stored in the directory. The
super user account of the directory should be used rarely and delegation of
privileges should be easier.

We think this proposed layout accomplishes that by providing several groups
which have distinctive access rules, providing a clear separation of
privileges. In order to give an user a new privilege, all is needed is to add
him/her to one of these specific groups.

These are the characteristics of the proposed DIT:
- several groups for common services
- most access control rules based on group membership
- several system accounts ready to use (just add a password) by many services
  such as:
  - sudo
  - dns
  - samba
  - etc
- simple installation script which prepares the tree asking very few questions
  (just two, and one of them is just a password)
- easy support for OpenLDAP's password policy overlay

These accounts get their privileges by being associated to specific group(s).

Administrators should note that we will probably find out that there are too
few groups, or too many. Or that some ACLs are too restrictive, or too broad.
It is difficult to come up with a one-size-fits-all DIT, but we can start here.

By the way, there is no password set for the "rootdn" account as it (the
account) is not used.

If you just want to know how to use this DIT, skip to the end of the document
to the section called "Enough with the theory: how to use this?".


The Tree
========

                             dc=example,dc=com

  ou=Hosts                        ou=System Groups            ou=System Accounts
  ou=Idmap                          cn=LDAP Admins              uid=Ldap Admin
  ou=Address Book                   cn=Sudo Admins              uid=Sudo Admin
  ou=dhcp                           cn=DNS Admins               uid=DNS Admin
  ou=dns                            cn=DNS Readers              uid=DNS Reader
  ou=People                         cn=DHCP Admins              uid=DHCP Admin
  ou=Group                          cn=Address Book Admins      uid=Address Book Admin
  ou=Password Policies              cn=LDAP Replicators         uid=LDAP Replicator
  ou=Sudoers                        cn=Account Admins           uid=Account Admin
                                    cn=MTA Admins               uid=MTA Admin
                                    cn=LDAP Monitors            uid=LDAP Monitor
                                    cn=Idmap Admins             uid=Idmap Admin
                                                                uid=smbldap-tools
                                                                uid=nssldap

The services
============

We created some entries for a few services that can use LDAP to store their
information. More will probably be added in the future. For now, we have
branches for:
- dns (ou=dns)
- sudo (ou=sudoers)
- dhcp (ou=dhcp)

The respective administrative groups have read/write access to these branches
for specific entries.


The groups
==========

Groups are the core of this proposed DIT layout, because most ACLs are
constructed via group membership to allow for greater flexibility and
delegation.

The current default groups that are born with the new DIT layout are as
follows:
- LDAP Admins
- Sudo Admins
- DNS Admins
- DNS Readers
- DHCP Admins
- Address Book Admins
- LDAP Replicators
- Account Admins
- MTA Admins
- LDAP Monitors
- Idmap Admins

Each entry has a description attribute filled in with a brief text describing
the purpose of the members of each group. For example:

dn: cn=Sudo Admins,ou=System Groups,dc=example,dc=com
description: Members can administer ou=sudoers entries and attributes

In order to use groups in ACLs, the objectClass used for these entries has to
use attributes where membership is indicated distinguished names and not just
names. In other words, the membership attribute has to use a full DN to
indicate its member. The standard object class used for this by OpenLDAP is
groupOfNames, and this is what we used. For example:

dn: cn=Sudo Admins,ou=System Groups,dc=example,dc=com
member: uid=Sudo Admin,ou=System Accounts,dc=example,dc=com

A side effect of using groupOfNames is that we *have* to have at least one
member in each group. So we needed to create standard accounts, which proved to
be usefull anyway. The previous example showed the standard account for
adminstering sudo entries and attributes.


The accounts
============

As was the case with the groups, many standard system accounts were created.
Each group has at least a corresponding system account as its membership. The
current list is as follows:

- Account Admin
- smbldap-tools
- nssldap
- MTA Admin
- DHCP Admin
- DNS Admin
- DNS Reader
- Sudo Admin
- Address Book Admin
- LDAP Admin
- LDAP Replicator
- LDAP Monitor
- Idmap Admin


The privileges
==============

The idea is to give each group the needed privileges to complete its
administration tasks. This usually means having access to the respective ou=foo
branch of the directory. For example, the Sudo Admins group has rights over the
ou=sudoers branch of the directory. 

Whenever possible, however, these rights are limited to that specific service,
i.e., it's not any kind of entry that can be created but just those relevant to
the service. For example, the Sudo Admins members can only create entries one
level below ou=sudoers, and only with the attributes allowed by the sudoRole
object class.

Other cases, however, are more complicated. We will list them here and the
reasoning behind the chosen ACLs.


Monitoring access
-----------------
The "LDAP Monitors" group is the only grop besides "LDAP Admins" which can read
entries under cn=monitor. This base dn contains statistics about the server,
such as operations performed, backends and overlays being used, etc. So, if you
need an user to have read access to this kind of information, just put him/her
in this group.


Samba, Unix and Kerberos admins
-------------------------------
Samba needs to have corresponding unix accounts for its users and machine
accounts. It will not by itself create those, however. For example, when
running "smbpasswd -a foo", the "foo" user account will only be created if
samba can find the corresponding unix attributes.  The same for group mappings
and machine accounts.

Earlier versions of openldap-dit had two separate privilege groups:
one for Unix accounts and another for Samba accounts. This complicated ACLs,
and it was worse when we later added Kerberos Admins to the mix because they
also had to touch some of the account-related attributes.

So, since version 0.11, we merged these groups into one called Account Admins
(and the respective Account Admin account). This made the ACLs simplier and
faster, at the expense of some granularity in privileges.

The smbldap-tools account, uid=smbldap-tools,ou=System Accounts, still exists
but is now a member of the Account Admins group.


MTA
---
As of this moment, there is no clear scenario for usage of this account. For
now, it can administer just a few attributes: all the ones from the
inetLocalMailRecipient object class plus the single mail attribute.

As more usage scenarios appear, these ACLs should be incremented.


DNS Readers
-----------
Members of this group are allowed read access to all attributes of the dNSZone
object class under ou=dns. Besides them and the members of the DNS Admins
group, no other entity can read these entries. This was done so to avoid the
"zone transfer" vulnerability scenario, where anonymous users could gather the
whole DNS database.


LDAP Admins
-----------
Members of this group can write to and read from all entries and attributes of
the directory and have no size or time limits.


LDAP Replicators
----------------
The members of the LDAP Replicators group have read access to all attributes
and entries of the directory so that they can be used in a syncrepl replication
setup. The bind dn used for the replication should be a member of this group.
For example:

syncrepl    rid=100
            provider=ldap://dirserv.example.com
            type=refreshAndPersist
            retry="60 +"
            searchbase="dc=example,dc=com"
            starttls=critical
            bindmethod=simple
            binddn="uid=LDAP Replicator,ou=System Accounts,dc=example,dc=com"
            credentials="secret"
		
Here, "uid=LDAP Replicator,ou=System Accounts,dc=example,dc=com" is a member of
the "LDAP Replicators" group and is automatically granted read rights to all
entries of the directory (assuming the provider was also installed with this
base DIT and ACLs).


Generic directory read accounts
-------------------------------
A few accounts were created for specific read access. Some administrators
prefer to block anonymous read access to the directory, in which case these
accounts would then be used. For the moment we have:
- nssldap: nss_ldap can bind to the directory either anonymously or with a
  specific account. The "uid=nssldap,ou=System Accounts" was created for this
  purpose.  Currently no ACLs make use of this account. Were the administrator to
  use it, he/she would also have to block anonymous read access to many
  attributes.

Currently anonymous read access is granted to many attributes. As of this
moment, if the administrator wants to restrict anonymous access and use these
accounts, the ACLs would have to be changed manually.


The installation script
=======================

The openldap-dit package contains a shell script which can be used to
install the accounts and ACLs described in this document. The script is
installed at /usr/share/openldap/scripts/openldap-dit-setup.sh and performs the
following:
- asks the DNS domain (suggesting whatever was auto-detected)
- constructs the top-level directory entry from this domain using dc style
  attributes
- creates and imports an ldif file with the accounts and groups described here
- installs new slapd.conf and openldap-dit-access.conf files (making backups of
  the previous ones) with the default ACLs and other useful configurations
  (like cache)
- loads the ldif file, backing up the previous database directory

Even though the script performs many tests and backups many files before
overwriting them, administrators are advised to backup all data before running
this script.


Enough with the theory: how to use this?
========================================

The installation script will overwrite some OpenLDAP files and directories.
Specifically, it will backup and overwrite the following:
- /etc/openldap/slapd.conf
- /etc/openldap/ldap.conf
- /etc/openldap/openldap-dit-access.conf (THIS ONE HAS NO BACKUP CURRENTLY)
- /var/lib/ldap contents

So, after you are satisfied that nothing important will be lost, run the
script. Below is a sample run using the example.com domain:

[root@cs4 ~]# /usr/share/openldap/scripts/openldap-dit-setup.sh
Please enter your DNS domain name [conectiva]:
example.com


Administrator account

The administrator account for this directory is
uid=LDAP Admin,ou=System Accounts,dc=example,dc=com

Please choose a password for this account:
New password:
Re-enter new password:


Summary
=======

Domain:      example.com
LDAP suffix: dc=example,dc=com

Confirm? (Y/n)
Y
config file testing succeeded
Stopping ldap service
Finished, starting ldap service
Running /usr/bin/db_recover on /var/lib/ldap
removing /var/lib/ldap/alock
Starting slapd (ldap + ldaps):                                  [  OK  ]

Your previous database directory has been backed up as /var/lib/ldap.1145397294


Now, fire up an LDAP browser and use the LDAP Admin account shown above to set
up some passwords for the other less privileged accounts that you are going to
use. Note that the "rootdn" account is not used.



debian/changelog
openldap-dit (0.19-1) unstable; urgency=low

  * Initial release 

 -- Andreas Hasenack <andreas@canonical.com>  Wed, 04 Jun 2008 17:02:22 -0300



kderc2ldif.sh
#!/bin/bash

# License: GPL
# Copyright: Mandriva (c) 2007

# This script will read a given KDE "rc" file and
# write its ldif representation to stdout according
# to kde.schema. See http://bugs.kde.org/show_bug.cgi?id=101716
# Example: ./kderc2ldif.sh konquerorrc

BASE="dc=example,dc=com"
KDEBASE="ou=KDEConfig,$BASE"
PROFILE="default"

function usage() {
	echo "Usage: $0 <kderc file>"
}


if [ "$#" -ne "1" ]; then
	usage
	exit 1
fi

file="$1"

if [ ! -e "$file" ]; then
	echo "File \"$file\" not found."
	exit 1
fi


cat <<EOF
dn: cn=$file,ou=$PROFILE,$KDEBASE
cn: $file
objectClass: appConfig
description: KDE configuration for $file
EOF


{ while read line; do
	# skip empty lines or lone [$i]
	if [ -z "$line" ]; then
		continue
	fi
	if [ "$line" = "[\$i]" -a -z "$globalimmutable" ]; then
		echo "immutable: TRUE"
		globalimmutable="true"
		continue
	fi
	# look for section
	if [ "${line:0:1}" = "[" ]; then
		if [ "${line:$((${#line}-4))}" = "[\$i]" -a -z "$globalimmutable" ]; then
			immutable="true"
		fi
		line="${line%\[\$i\]}"
		temp=${line:1}
		section=${temp/]/}
		echo ""
		echo "dn: cn=$section,cn=$file,ou=$PROFILE,$KDEBASE"
		echo "cn: $section"
		echo "objectClass: appConfigSection"
		echo "appName: $file"
		if [ "$immutable" = "true" ]; then
			echo "immuteable: TRUE"
		fi
	elif [ -n "$section" ]; then
		echo "appConfigEntry: $line"
	fi
done } < "$file"
echo




mandriva-dit-setup.sh
#!/bin/bash

if [ "`id -u`" != "0" ]; then
	echo "Error, must be root user"
	exit 1
fi

SLAPTEST="/usr/sbin/slaptest"
SLAPADD="/usr/sbin/slapadd"
SLAPPASSWD="/usr/sbin/slappasswd"
slapd_conf_template="/usr/share/openldap/mandriva-dit/mandriva-dit-slapd-template.conf"
base_ldif_template="/usr/share/openldap/mandriva-dit/mandriva-dit-base-template.ldif"
acl_template="/usr/share/openldap/mandriva-dit/mandriva-dit-access-template.conf"
acl_file="/etc/openldap/mandriva-dit-access.conf"
now=`date +%s`
myfqdn=`hostname -f`
verbose=
noprompt=
if [ -z "$myfqdn" ]; then
	myfqdn="localhost"
fi

function usage() {
	echo "Usage:"
	echo "$0 [-h | --help] [-v] [-d <dnsdomain>] [-p <password>] [-y]"
	echo
	echo "-h | --help    : shows this help"
	echo "-v             : verbose mode"
	echo "-d <dnsdomain> : use <dnsdomain> for dns domain"
	echo "-p <password>  : use <password> for LDAP Admin password"
	echo 
	echo "-y             : assume default answer in all prompts "
	echo "                 except the password one"
}

function echo_v() {
	if [ -n "$verbose" ]; then
		echo "== $@"
	fi
}

# output: stdout: example.com or the possible detected domain
function detect_domain() {
	mydomain=`hostname -d`
	if [ -z "$mydomain" ]; then
		mydomain="example.com"
	fi
	echo "$mydomain"
	return 0
}

# output: stdout: name of temporary file
# *exits* if error
function make_temp() {
	tmpfile=`mktemp ${TMP:-/tmp}/mandriva-dit.XXXXXXXXXXXX`
	if [ -f "$tmpfile" ]; then
		echo "$tmpfile"
		return 0
	else
		echo "error while making temporary file" >&2
		exit 1
	fi
}

# $1: domain
# returns standard dc=foo,dc=bar suffix on stdout
function calc_suffix() {
	old_ifs=${IFS}
	IFS="."
	for component in $1; do
		result="$result,dc=$component"
	done
	IFS="${old_ifs}"
	echo "${result#,}"
	return 0
}


# $1: directory where the LDAP database is
# output (stdout): backup dir of the previous database
function clean_database() {
	backupdir="/var/lib/ldap.$now"
	cp -a "$1" "$backupdir" 2>/dev/null
	if [ "$?" -ne "0" ]; then
		echo "Error, could not make a backup copy of the"
		echo "current LDAP database, aborting..."
		echo
		echo "(not enough disk space?)"
		echo
		exit 1
	fi
	rm -f "$1"/{*.bdb,log.*,__*,alock}
	echo "$backupdir"
	return 0
}

# $1: *file* to be backed up
# output (stdout): backup filename
function mybackup() {
	if [ ! -f "$1" ]; then
		echo "Internal error, $1 has to be a file" >&2
		echo "Aborting" >&2
		exit 1
	fi
	newname="$1.$now"
	cp -a "$1" "$newname"
	echo "$newname"
	return 0
}

while [ -n "$1" ]; do
	case "$1" in
		-h | --help)
		usage
		exit 1
		;;
		-v)
		verbose=1
		shift
		;;
		-d)
		shift
		if [ -n "$1" -a "${1##-}" != "${1}" -o -z "${1}" ]; then
			echo "Error, -d requires an argument"
			exit 1
		fi
		mydomain="$1"
		shift
		;;
		-p)
		shift
		if [ -n "$1" -a "${1##-}" != "${1}" -o -z "${1}" ]; then
			echo "Error, -p requires an argument"
			exit 1
		fi
		mypass="$1"
		shift
		;;
		-y)
		noprompt=1
		shift
		;;
	esac
done

echo_v
echo_v "Running in verbose mode"
echo_v

if [ -z "$mydomain" ]; then
	mydomain=`detect_domain`
	if [ -z "$noprompt" ]; then
		echo "Please enter your DNS domain name [$mydomain]:"
		read inputdomain
		if [ -n "$inputdomain" ]; then
			mydomain="$inputdomain"
		fi
	fi
fi
mysuffix=`calc_suffix $mydomain`

if [ -z "$mypass" ]; then
	echo
	echo "Administrator account"
	echo
	echo "The administrator account for this directory is"
	echo "uid=LDAP Admin,ou=System Accounts,$mysuffix"
	echo
	echo "Please choose a password for this account:"
	while /bin/true; do
		pass=`$SLAPPASSWD -h {SSHA}`
		if [ -n "$pass" ]; then
			break
		fi
	done
else
	pass=`$SLAPPASSWD -h {SSHA} -s "$mypass"`
	if [ "$?" -ne "0" ]; then
		echo "Error at generating password, exiting"
		exit 1
	fi
fi

# confirmation
echo
echo
echo "Summary"
echo "======="
echo
echo "Domain:        $mydomain"
echo "LDAP suffix:   $mysuffix"
echo "Administrator: uid=LDAP Admin,ou=System Accounts,$mysuffix"
echo
if [ -z "$noprompt" ]; then
	echo "Confirm? (Y/n)"
	read val
	if [ "$val" = "n" -o "$val" = "N" ]; then
		echo
		echo "Cancelled."
		exit 1
	fi
fi

# let's do it

# first, slapd.conf
myslapdconf=`make_temp`

echo_v "Creating slapd.conf..."
cat $slapd_conf_template | sed -e "s/@SUFFIX@/$mysuffix/g;" > $myslapdconf
chmod 0640 $myslapdconf
chgrp ldap $myslapdconf

# now, /etc/openldap/ldap.conf
myldapconf=`make_temp`
echo_v "Creating ldap.conf..."
cat /etc/openldap/ldap.conf | \
	sed -e "s/^BASE[[:blank:]]\+.*/BASE $mysuffix/g;\
	s/^HOST[[:blank:]]\+.*/HOST $myfqdn/g;\
	s@^URI[[:blank:]]\+.*@URI ldap://$myfqdn@g" \
	> $myldapconf
if ! grep -qE '^(HOST|URI)' $myldapconf; then
	echo "URI ldap://$myfqdn" >> $myldapconf
fi
if ! grep -qE '^BASE' $myldapconf; then
	echo "BASE $mysuffix" >> $myldapconf
fi
chmod 0644 $myldapconf

# ACL file
echo_v "Creating acl file..."
cat $acl_template | sed -e "s/@SUFFIX@/$mysuffix/g" > $acl_file
chmod 0640 $acl_file
chgrp ldap $acl_file

# test
echo_v "Testing configuration with temporary files..."
$SLAPTEST -u -f $myslapdconf
if [ "$?" -ne "0" ]; then
	echo "ERROR"
	echo "OpenLDAP configuration file was generated with errors"
	echo "Aborting. File used was $myslapdconf"
	rm -f $myldapconf
	exit 1
fi

# load
echo_v "Test succeeded, creating LDIF file for database loading..."
myldif=`make_temp`
cat $base_ldif_template | sed -e "\
	s/@SUFFIX@/$mysuffix/g;\
	s/@DC@/${mydomain%%.[a-zA-Z0-9]*}/g;\
	s/@DOMAIN@/${mydomain}/g;\
	s|@ldapadmin_password@|$pass|g" > $myldif

# dry run first
echo_v "Attempting to test-load the database..."
$SLAPADD -u -f $myslapdconf < $myldif
if [ "$?" -ne "0" ]; then
	echo "ERROR"
	echo "Database loading failed during test run."
	echo "Ldif file used:       $myldif"
	echo "slapd.conf file used: $myslapdconf"
	echo
	echo "Exiting"
	rm -f $myldapconf
	exit 1
fi

# let's go for real now
echo_v "Everything ok, doing it for real now."
echo "Stopping ldap service"
/sbin/service ldap stop > /dev/null 2>&1 || :
echo_v "Backing up existing files..."
backup_db=`clean_database /var/lib/ldap`
backup_slapd_conf=`mybackup /etc/openldap/slapd.conf`
backup_ldap_conf=`mybackup /etc/openldap/ldap.conf`
echo_v "Writing /etc/openldap/slapd.conf and /etc/openldap/ldap.conf..."
cat $myslapdconf > /etc/openldap/slapd.conf; rm -f $myslapdconf
cat $myldapconf > /etc/openldap/ldap.conf; rm -f $myldapconf

echo_v "Loading database..."
$SLAPADD < $myldif
if [ "$?" -ne "0" ]; then
	echo "Something went wrong while initializing the database"
	echo "Aborting. Your previous database is at $backup_db"
	echo "Your original /etc/openldap/{slapd,ldap}.conf files"
	echo "were backed up as $backup_slapd_conf and"
	echo "$backup_ldap_conf, respectively."
	exit 1
fi

echo "Finished, starting ldap service"
/sbin/service ldap start
echo
echo "Your previous database directory has been backed up as $backup_db"
echo "All files that were backed up got the suffix \"$now\"."
echo


openldap-dit-setup.sh
#!/bin/bash

if [ "`id -u`" != "0" ]; then
	echo "Error, must be root user"
	exit 1
fi

function distro_guess()
{
#$ cat /etc/lsb-release 
#DISTRIB_ID=Ubuntu
#DISTRIB_RELEASE=8.04
#DISTRIB_CODENAME=hardy
#DISTRIB_DESCRIPTION="Ubuntu 8.04"
    if [ -r "/etc/lsb-release" ]; then
        source /etc/lsb-release
    else
        echo "Can't guess distro name (no /etc/lsb-release or it's not readable)"
        exit 1
    fi
    if [ -z "$DISTRIB_ID" -o -z "$DISTRIB_RELEASE" ]; then
        echo "No DISTRIB_ID or DISTRIB_RELEASE variable(s) in /etc/lsb-release"
        exit 1
    fi
    DISTRIB_ID=`echo $DISTRIB_ID | tr A-Z a-z`
    export DISTRIB_ID DISTRIB_RELEASE
    echo $DISTRIB_ID
    return 0
}

function mandriva_setup()
{
    export SLAPTEST="/usr/sbin/slaptest"
    export SLAPADD="/usr/sbin/slapadd"
    export SLAPPASSWD="/usr/sbin/slappasswd"
    export SERVICE="/sbin/service ldap"
    export slapd_conf_template="/usr/share/openldap/openldap-dit/openldap-dit-slapd-template.conf"
    export ldap_conf_dir="/etc/openldap"
    export slapd_conf="$ldap_conf_dir/slapd.conf"
    export schema_dir="/usr/share/openldap/schema"
    export ldap_libdir="/usr/lib/openldap"
    export ldap_conf="$ldap_conf_dir/ldap.conf"
    export base_ldif_template="/usr/share/openldap/openldap-dit/openldap-dit-base-template.ldif"
    export acl_template="/usr/share/openldap/openldap-dit/openldap-dit-access-template.conf"
    export acl_file="$ldap_conf_dir/openldap-dit-access.conf"
    export run_dir="/var/run/ldap"
    export db_dir="/var/lib/ldap"
    export ldap_user="ldap"
    export ldap_group="ldap"

    for package in openldap-server openldap-client; do
        if ! rpm -q $package 2>/dev/null; then
            echo "Error, please install package $package"
        fi
    done

    return 0
}

function ubuntu_setup()
{
    export SLAPTEST="/usr/sbin/slaptest"
    export SLAPADD="/usr/sbin/slapadd"
    export SLAPPASSWD="/usr/sbin/slappasswd"
    if [ -x /usr/sbin/invoke-rc.d ]; then
        SERVICE="/usr/sbin/invoke-rc.d slapd"
    else
        SERVICE="/etc/init.d/slapd"
    fi
    export slapd_conf_template="/usr/share/slapd/openldap-dit/openldap-dit-slapd-template.conf"
    export ldap_conf_dir="/etc/ldap"
    export slapd_conf="$ldap_conf_dir/slapd.conf"
    export schema_dir="$ldap_conf_dir/schema"
    export ldap_libdir="/usr/lib/ldap"
    export ldap_conf="$ldap_conf_dir/ldap.conf"
    export base_ldif_template="/usr/share/slapd/openldap-dit/openldap-dit-base-template.ldif"
    export acl_template="/usr/share/slapd/openldap-dit/openldap-dit-access-template.conf"
    export acl_file="$ldap_conf_dir/openldap-dit-access.conf"
    export run_dir="/var/run/slapd"
    export db_dir="/var/lib/ldap"
    export ldap_user="openldap"
    export ldap_group="openldap"

    mkdir -p `dirname $slapd_conf_template`

    for package in slapd ldap-utils; do
        if ! dpkg -l $package 2>/dev/null | grep -q ^ii; then
            echo "Error, please install package $package"
            exit 1
        fi
    done

    return 0
}

now=`date +%s`
myfqdn=`hostname -f`
verbose=
noprompt=
if [ -z "$myfqdn" ]; then
	myfqdn="localhost"
fi
distro=`distro_guess`
${distro}_setup

function usage() {
	echo "Usage:"
	echo "$0 [-h | --help] [-v] [-d <dnsdomain>] [-p <password>] [-y]"
	echo
	echo "-h | --help    : shows this help"
	echo "-v             : verbose mode"
	echo "-d <dnsdomain> : use <dnsdomain> for dns domain"
	echo "-p <password>  : use <password> for LDAP Admin password"
	echo 
	echo "-y             : assume default answer in all prompts "
	echo "                 except the password one"
}

function echo_v() {
	if [ -n "$verbose" ]; then
		echo "== $@"
	fi
}

# output: stdout: example.com or the possible detected domain
function detect_domain() {
	mydomain=`hostname -d`
	if [ -z "$mydomain" ]; then
		mydomain="example.com"
	fi
	echo "$mydomain"
	return 0
}

# output: stdout: name of temporary file
# *exits* if error
function make_temp() {
	tmpfile=`mktemp ${TMP:-/tmp}/openldap-dit.XXXXXXXXXXXX`
	if [ -f "$tmpfile" ]; then
		echo "$tmpfile"
		return 0
	else
		echo "error while making temporary file" >&2
		exit 1
	fi
}

# $1: domain
# returns standard dc=foo,dc=bar suffix on stdout
function calc_suffix() {
	old_ifs=${IFS}
	IFS="."
	for component in $1; do
		result="$result,dc=$component"
	done
	IFS="${old_ifs}"
	echo "${result#,}"
	return 0
}


# $1: directory where the LDAP database is
# output (stdout): backup dir of the previous database
function clean_database() {
	backupdir="$db_dir.$now"
	cp -a "$1" "$backupdir" 2>/dev/null
	if [ "$?" -ne "0" ]; then
		echo "Error, could not make a backup copy of the"
		echo "current LDAP database, aborting..."
		echo
		echo "(not enough disk space?)"
		echo
		exit 1
	fi
	rm -f "$1"/{*.bdb,log.*,__*,alock}
	echo "$backupdir"
	return 0
}

# $1: *file* to be backed up
# output (stdout): backup filename
function mybackup() {
	if [ ! -f "$1" ]; then
		echo "Internal error, $1 has to be a file" >&2
		echo "Aborting" >&2
		exit 1
	fi
	newname="$1.$now"
	cp -a "$1" "$newname"
	echo "$newname"
	return 0
}

while [ -n "$1" ]; do
	case "$1" in
		-h | --help)
		usage
		exit 1
		;;
		-v)
		verbose=1
		shift
		;;
		-d)
		shift
		if [ -n "$1" -a "${1##-}" != "${1}" -o -z "${1}" ]; then
			echo "Error, -d requires an argument"
			exit 1
		fi
		mydomain="$1"
		shift
		;;
		-p)
		shift
		if [ -n "$1" -a "${1##-}" != "${1}" -o -z "${1}" ]; then
			echo "Error, -p requires an argument"
			exit 1
		fi
		mypass="$1"
		shift
		;;
		-y)
		noprompt=1
		shift
		;;
	esac
done

echo_v
echo_v "Running in verbose mode"
echo_v

if [ -z "$mydomain" ]; then
	mydomain=`detect_domain`
	if [ -z "$noprompt" ]; then
		echo "Please enter your DNS domain name [$mydomain]:"
		read inputdomain
		if [ -n "$inputdomain" ]; then
			mydomain="$inputdomain"
		fi
	fi
fi
mysuffix=`calc_suffix $mydomain`

if [ -z "$mypass" ]; then
	echo
	echo "Administrator account"
	echo
	echo "The administrator account for this directory is"
	echo "uid=LDAP Admin,ou=System Accounts,$mysuffix"
	echo
	echo "Please choose a password for this account:"
	while /bin/true; do
		pass=`$SLAPPASSWD -h {SSHA}`
		if [ -n "$pass" ]; then
			break
		fi
	done
else
	pass=`$SLAPPASSWD -h {SSHA} -s "$mypass"`
	if [ "$?" -ne "0" ]; then
		echo "Error at generating password, exiting"
		exit 1
	fi
fi

# confirmation
echo
echo
echo "Summary"
echo "======="
echo
echo "Domain:        $mydomain"
echo "LDAP suffix:   $mysuffix"
echo "Administrator: uid=LDAP Admin,ou=System Accounts,$mysuffix"
echo
if [ -z "$noprompt" ]; then
	echo "Confirm? (Y/n)"
	read val
	if [ "$val" = "n" -o "$val" = "N" ]; then
		echo
		echo "Cancelled."
		exit 1
	fi
fi

# let's do it

# first, slapd.conf
myslapdconf=`make_temp`

echo_v "Creating slapd.conf..."
cat $slapd_conf_template | sed -e "\
    s/@SUFFIX@/$mysuffix/g;\
    s,@SCHEMADIR@,$schema_dir,g;\
    s,@LDAPCONFDIR@,$ldap_conf_dir,g;\
    s,@RUNDIR@,$run_dir,g;\
    s,@LDAPLIBDIR@,$ldap_libdir,g;" > $myslapdconf
chmod 0640 $myslapdconf
chgrp $ldap_group $myslapdconf

# now, /etc/openldap/ldap.conf
myldapconf=`make_temp`
echo_v "Creating $ldap_conf..."
cat $ldap_conf | \
	sed -e "s/^BASE[[:blank:]]\+.*/BASE $mysuffix/g;\
	s/^HOST[[:blank:]]\+.*/HOST $myfqdn/g;\
	s@^URI[[:blank:]]\+.*@URI ldap://$myfqdn@g" \
	> $myldapconf
if ! grep -qE '^(HOST|URI)' $myldapconf; then
	echo "URI ldap://$myfqdn" >> $myldapconf
fi
if ! grep -qE '^BASE' $myldapconf; then
	echo "BASE $mysuffix" >> $myldapconf
fi
chmod 0644 $myldapconf

# ACL file
echo_v "Creating acl file..."
cat $acl_template | sed -e "s/@SUFFIX@/$mysuffix/g" > $acl_file
chmod 0640 $acl_file
chgrp $ldap_group $acl_file

# test
echo_v "Testing configuration with temporary files..."
$SLAPTEST -u -f $myslapdconf
if [ "$?" -ne "0" ]; then
	echo "ERROR"
	echo "OpenLDAP configuration file was generated with errors"
	echo "Aborting. File used was $myslapdconf"
	rm -f $myldapconf
	exit 1
fi

# load
echo_v "Test succeeded, creating LDIF file for database loading..."
myldif=`make_temp`
cat $base_ldif_template | sed -e "\
	s/@SUFFIX@/$mysuffix/g;\
	s/@DC@/${mydomain%%.[a-zA-Z0-9]*}/g;\
	s/@DOMAIN@/${mydomain}/g;\
	s|@ldapadmin_password@|$pass|g" > $myldif

# dry run first
echo_v "Attempting to test-load the database..."
$SLAPADD -u -f $myslapdconf < $myldif
if [ "$?" -ne "0" ]; then
	echo "ERROR"
	echo "Database loading failed during test run."
	echo "Ldif file used:       $myldif"
	echo "slapd.conf file used: $myslapdconf"
	echo
	echo "Exiting"
	rm -f $myldapconf
	exit 1
fi

# let's go for real now
echo_v "Everything ok, doing it for real now."
echo "Stopping ldap service"
$SERVICE stop > /dev/null 2>&1 || :
echo_v "Backing up existing files..."
backup_db=`clean_database $db_dir`
backup_slapd_conf=`mybackup $slapd_conf`
backup_ldap_conf=`mybackup $ldap_conf`
echo_v "Writing $slapd_conf and $ldap_conf..."
cat $myslapdconf > $slapd_conf; rm -f $myslapdconf
cat $myldapconf > $ldap_conf; rm -f $myldapconf

echo_v "Loading database..."
$SLAPADD < $myldif
if [ "$?" -ne "0" ]; then
	echo "Something went wrong while initializing the database"
	echo "Aborting. Your previous database is at $backup_db"
	echo "Your original slapd.conf and ldap.conf files"
	echo "were backed up as $backup_slapd_conf and"
	echo "$backup_ldap_conf, respectively."
	exit 1
fi

# forcibly adjust ownership of the DB files, as the Debian/Ubuntu
# initscript doesn't do it automatically
chown -R $ldap_user:$ldap_group $db_dir/*

echo "Finished, starting ldap service"
$SERVICE start
echo
echo "Your previous database directory has been backed up as $backup_db"
echo "All files that were backed up got the suffix \"$now\"."
echo
